### PR TITLE
Use Cluster Proxy for cluster communication

### PR DESCRIPTION
**Pull request by keyki**

### CODE REVIEW

File: core/src/main/resources/schema/app/20191128131307_CB-4356_alter_stack_add_column_tunnel.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+ALTER TABLE stack ADD COLUMN IF NOT EXISTS tunnel VARCHAR(255);
+UPDATE stack SET tunnel='CCM' WHERE useccm=true;
+UPDATE stack SET tunnel='DIRECT' WHERE useccm is null OR useccm=false;
+
+-- //@UNDO
+-- SQL to undo the change goes here.
+
+ALTER TABLE stack DROP COLUMN IF EXISTS tunnel;
+UPDATE stack SET useccm=true WHERE tunnel='CCM';
```
Comment (topolyai5) [inline core/src/main/resources/schema/app/20191128131307_CB-4356_alter_stack_add_column_tunnel.sql]: wrong order of sql

File: core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/clusterproxy/ClusterProxyServiceTest.java [hunk 2]
```diff
@@ -37,12 +37,14 @@
...
 import com.sequenceiq.cloudbreak.service.secret.domain.Secret;
 import com.sequenceiq.cloudbreak.service.secret.vault.VaultConfigException;
 import com.sequenceiq.cloudbreak.service.secret.vault.VaultSecret;
 import com.sequenceiq.cloudbreak.service.securityconfig.SecurityConfigService;
 import com.sequenceiq.cloudbreak.service.stack.StackService;
 import com.sequenceiq.common.api.type.InstanceGroupType;
+import com.sequenceiq.common.api.type.Tunnel;
 
 @RunWith(MockitoJUnitRunner.class)
 public class ClusterProxyServiceTest {
```
Comment (topolyai5) [inline core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/clusterproxy/ClusterProxyServiceTest.java]: additional unit tests would be good

File: environment-api/src/main/java/com/sequenceiq/environment/api/doc/environment/EnvironmentModelDescription.java [hunk 3]
```diff
@@ -50,7 +50,8 @@
     public static final String DEFAULT_SECURITY_GROUP = "Security group where all other hosts are placed.";
     public static final String SECURITY_CIDR = "CIDR range which is allowed for inbound traffic. Either IPv4 or IPv6 is allowed.";
 
-    public static final String TUNNEL = "Configuration that the connection going directly or with ccm.";
+    public static final String TUNNEL = "Configuration that the connection going directly or with cluster proxy or with ccm and cluster proxy.";
+    public static final String CLUSTER_PROXY = "Configuration that the connection going directly or with cluster proxy. Must be ";
```
Comment (topolyai5) [inline environment-api/src/main/java/com/sequenceiq/environment/api/doc/environment/EnvironmentModelDescription.java]: Must be what?

File: freeipa/src/main/resources/schema/app/20191128131307_CB-4356_alter_stack_add_column_tunnel.sql [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE stack ADD COLUMN IF NOT EXISTS tunnel VARCHAR(255);
+UPDATE stack SET tunnel='CCM' WHERE useccm=true;
+UPDATE stack SET tunnel='DIRECT' WHERE useccm is null OR useccm=false;
+
+-- //@UNDO
+-- SQL to undo the change goes here.
+
+ALTER TABLE stack DROP COLUMN IF EXISTS tunnel;
```
Comment (topolyai5) [inline freeipa/src/main/resources/schema/app/20191128131307_CB-4356_alter_stack_add_column_tunnel.sql]: wrong sql order

File: common/src/main/java/com/sequenceiq/cloudbreak/client/DisableProxyAuthFilter.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import javax.ws.rs.client.ClientRequestFilter;
+import javax.ws.rs.core.MultivaluedMap;
+
+public class DisableProxyAuthFilter implements ClientRequestFilter {
+    @Override
+    public void filter(ClientRequestContext requestContext) {
+        MultivaluedMap<String, Object> headers = requestContext.getHeaders();
+        headers.add("Proxy-Ignore-Auth", "true");
+    }
+}
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply common/src/main/java/com/sequenceiq/cloudbreak/client/DisableProxyAuthFilter.java]: yes

File: client-cm/src/main/java/com/sequenceiq/cloudbreak/cm/client/ClouderaManagerApiClientProvider.java [hunk 6]
```diff
@@ -84,7 +84,7 @@ private ApiClient decorateClient(HttpClientConfig clientConfig, String userName,
         cmClient.setVerifyingSsl(true);
 
         try {
-            if (isCmSslConfigValidClientConfigValid(clientConfig)) {
+            if (isCmSslConfigValidClientConfigValid(clientConfig) && !clientConfig.isClusterProxyenabled()) {
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply client-cm/src/main/java/com/sequenceiq/cloudbreak/cm/client/ClouderaManagerApiClientProvider.java]: fixed

File: cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerSecurityService.java [hunk 7]
```diff
@@ -255,7 +255,11 @@ private void createNewUser(UsersResourceApi usersResourceApi, List<ApiAuthRoleRe
     }
 
     public ApiClient getDefaultClient(Integer gatewayPort, HttpClientConfig clientConfig) throws ClouderaManagerClientInitException {
-        return clouderaManagerApiClientProvider.getClouderaManagerClient(clientConfig, gatewayPort, "admin", "admin");
+        ApiClient client = clouderaManagerApiClientProvider.getClouderaManagerClient(clientConfig, gatewayPort, "admin", "admin");
+        if (clientConfig.isClusterProxyenabled()) {
+            client.addDefaultHeader("Proxy-Ignore-Auth", "true");
+        }
+        return client;
     }
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerSecurityService.java]: sure, fixed

File: common/src/main/java/com/sequenceiq/cloudbreak/auth/ThreadBasedUserCrnProvider.java [hunk 8]
```diff
@@ -27,7 +27,7 @@ public String getUserCrn() {
     public String getAccountId() {
         String userCrn = getUserCrn();
         if (userCrn != null) {
-            return Optional.ofNullable(Crn.fromString(userCrn)).orElseThrow(() -> new IllegalStateException("Unable to obtain crn!")).getAccountId();
+            return Optional.ofNullable(Crn.fromString(userCrn)).orElseThrow(() ->   new IllegalStateException("Unable to obtain crn!")).getAccountId();
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply common/src/main/java/com/sequenceiq/cloudbreak/auth/ThreadBasedUserCrnProvider.java]: nope

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java [hunk 9]
```diff
@@ -49,15 +47,8 @@
...
-import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmParameterSupplier;
-import com.sequenceiq.cloudbreak.ccm.endpoint.KnownServiceIdentifier;
-import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceFamilies;
-import com.sequenceiq.cloudbreak.domain.projection.StackListItem;
-import com.sequenceiq.cloudbreak.telemetry.fluent.FluentClusterType;
-import com.sequenceiq.cloudbreak.telemetry.fluent.cloud.CloudStorageFolderResolverService;
-import com.sequenceiq.cloudbreak.workspace.authorization.PermissionCheckingUtils;
 import com.sequenceiq.cloudbreak.blueprint.validation.AmbariBlueprintValidator;
-import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmParameters;
+import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmParameterSupplier;
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java]: removed

File: core/src/main/resources/schema/app/20191128131307_CB-4356_alter_stack_add_column_tunnel.sql (hunk 1 above)
Comment (lacikaaa) [reply core/src/main/resources/schema/app/20191128131307_CB-4356_alter_stack_add_column_tunnel.sql]: good catch!

File: core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/clusterproxy/ClusterProxyServiceTest.java (hunk 2 above)
Comment (lacikaaa) [reply core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/clusterproxy/ClusterProxyServiceTest.java]: yeah...

File: environment-api/src/main/java/com/sequenceiq/environment/api/doc/environment/EnvironmentModelDescription.java (hunk 3 above)
Comment (lacikaaa) [reply environment-api/src/main/java/com/sequenceiq/environment/api/doc/environment/EnvironmentModelDescription.java]: removed, unused

File: freeipa/src/main/resources/schema/app/20191128131307_CB-4356_alter_stack_add_column_tunnel.sql (hunk 4 above)
Comment (lacikaaa) [reply freeipa/src/main/resources/schema/app/20191128131307_CB-4356_alter_stack_add_column_tunnel.sql]: good catch!

**Pull request merged**